### ISSUE TITLE
Possibly incorrect nullable annotation on IndexExpression.Update()

### ISSUE THREAD

**Issue by roji** [dotnet/runtime#96626]
The [Expression.Property](https://learn.microsoft.com/en-us/dotnet/api/system.linq.expressions.expression.property?view=net-8.0#system-linq-expressions-expression-property(system-linq-expressions-expression-system-reflection-propertyinfo-system-linq-expressions-expression())) API allows constructing an IndexExpression with a null object (representing an invocation of a static indexer property). However, [IndexExpression.Update()](https://learn.microsoft.com/en-us/dotnet/api/system.linq.expressions.indexexpression.update?view=net-8.0) is annotated to accept a non-nullable object parameter; it seems like this should be relaxed to accept a nullable one.

**Comment by stephentoub** [comment]
As it's currently written, IndexExpression.Update calls MakeIndex with that object:
https://github.com/dotnet/runtime/blob/623cf77a58f7a233b94dcc1c3ef8eb8d67e8d948/src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/IndexExpression.cs#L87-L97
MakeIndex might pass it to ArrayAccess:
https://github.com/dotnet/runtime/blob/623cf77a58f7a233b94dcc1c3ef8eb8d67e8d948/src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/IndexExpression.cs#L138-L146
and ArrayAccess will fail if the object is null:
https://github.com/dotnet/runtime/blob/623cf77a58f7a233b94dcc1c3ef8eb8d67e8d948/src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/IndexExpression.cs#L175-L177
We could still update the annotation, since null is permitted in at least some circumstances, but there might be a few desirable tweaks beyond just updating the annotation.

**Comment by roji** [comment]
Right, this is the result of IndexExpression being used to represent two different things (indexer property and array).
